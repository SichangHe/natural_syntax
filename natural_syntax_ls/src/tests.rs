use std::time::Instant;

use super::*;

use insta::assert_debug_snapshot;
use tracing::Level;
use tracing_subscriber::EnvFilter;

#[test]
fn token_types() {
    init_tracing();
    let types = semantic_token_types();
    assert_debug_snapshot!(types);
}

#[test]
fn token_modifiers() {
    init_tracing();
    let modifiers = semantic_token_modifiers();
    assert_debug_snapshot!(modifiers);
}

#[test]
fn filtering_tokens() {
    assert!(filter_token(&dummy_token("word", 1.)));
    assert!(filter_token(&dummy_token("word,", 0.4)));
    assert!(!filter_token(&dummy_token("word", 0.3)));
    assert!(!filter_token(&dummy_token(",", 0.3)));
    assert!(!filter_token(&dummy_token(",", 1.)));
    assert!(!filter_token(&dummy_token("]", 1.)));
    assert!(!filter_token(&dummy_token("(", 1.)));
}

fn dummy_token(word: &str, score: f64) -> POSToken {
    POSToken {
        word: word.into(),
        score,
        ..Default::default()
    }
}

#[test]
fn convert_tokens() {
    init_tracing();
    let start = Instant::now();
    let model = POSModel::try_default().unwrap();
    info!("Took {}ms to load the model.", start.elapsed().as_millis());
    let input = "The easiest way of getting started is the [rustler Elixir library](https://hex.pm/packages/rustler).

- Add the [rustler Elixir library](https://hex.pm/packages/rustler) as a
  dependency of your project.
- Run `mix rustler.new` to generate a new NIF in your project. Follow the
  instructions.
- If you are already using [`serde`](https://serde.rs) and/or have been using
  `serde_rustler` before, please enable the `serde` feature in your NIF crate's
  `Cargo.toml` on the `rustler` dependency.
";
    let start = Instant::now();
    let output = model.predict(input);
    info!("Took {}ms to predict.", start.elapsed().as_millis());
    let mut tokens = output
        .map(|r| r.unwrap())
        .filter(filter_token)
        .collect::<Vec<_>>();
    round_scores(&mut tokens);
    let text = Rope::from_str(input);
    let semantic_tokens = semantic_tokens(&text, &tokens, &Default::default());
    let combined = tokens.into_iter().zip(semantic_tokens).collect::<Vec<_>>();
    assert_debug_snapshot!(combined);
}

const PRECISION: f64 = 1e-3;

fn round_scores(predictions: &mut [POSToken]) {
    predictions
        .iter_mut()
        .for_each(|token| token.score = (token.score / PRECISION).round() * PRECISION)
}

fn init_tracing() {
    _ = tracing_subscriber::fmt()
        .with_env_filter(
            EnvFilter::builder()
                .with_default_directive(Level::INFO.into())
                .from_env_lossy(),
        )
        .try_init();
}
