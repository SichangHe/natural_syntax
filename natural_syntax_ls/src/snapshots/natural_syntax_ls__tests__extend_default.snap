---
source: natural_syntax_ls/src/tests.rs
expression: token_map
---
TokenMap(
    {
        FW: Some(
            TokenBits {
                token_type: 18,
                token_modifiers_bitset: 0,
            },
        ),
        NNP: Some(
            TokenBits {
                token_type: 7,
                token_modifiers_bitset: 1,
            },
        ),
        NNPS: Some(
            TokenBits {
                token_type: 7,
                token_modifiers_bitset: 129,
            },
        ),
        RBR: Some(
            TokenBits {
                token_type: 10,
                token_modifiers_bitset: 64,
            },
        ),
        RB: Some(
            TokenBits {
                token_type: 10,
                token_modifiers_bitset: 0,
            },
        ),
        SYM: Some(
            TokenBits {
                token_type: 21,
                token_modifiers_bitset: 256,
            },
        ),
        CD: Some(
            TokenBits {
                token_type: 19,
                token_modifiers_bitset: 0,
            },
        ),
        VBN: Some(
            TokenBits {
                token_type: 13,
                token_modifiers_bitset: 512,
            },
        ),
        WP: Some(
            TokenBits {
                token_type: 15,
                token_modifiers_bitset: 512,
            },
        ),
        IN: None,
        VB: Some(
            TokenBits {
                token_type: 12,
                token_modifiers_bitset: 0,
            },
        ),
        UH: Some(
            TokenBits {
                token_type: 15,
                token_modifiers_bitset: 128,
            },
        ),
        NN: Some(
            TokenBits {
                token_type: 7,
                token_modifiers_bitset: 0,
            },
        ),
        VBD: Some(
            TokenBits {
                token_type: 12,
                token_modifiers_bitset: 128,
            },
        ),
        RP: Some(
            TokenBits {
                token_type: 21,
                token_modifiers_bitset: 0,
            },
        ),
        JJ: Some(
            TokenBits {
                token_type: 1,
                token_modifiers_bitset: 0,
            },
        ),
        VBZ: Some(
            TokenBits {
                token_type: 13,
                token_modifiers_bitset: 8,
            },
        ),
        JJR: Some(
            TokenBits {
                token_type: 5,
                token_modifiers_bitset: 128,
            },
        ),
        MD: Some(
            TokenBits {
                token_type: 15,
                token_modifiers_bitset: 4,
            },
        ),
        TO: Some(
            TokenBits {
                token_type: 10,
                token_modifiers_bitset: 0,
            },
        ),
        POS: Some(
            TokenBits {
                token_type: 9,
                token_modifiers_bitset: 1,
            },
        ),
        NNS: Some(
            TokenBits {
                token_type: 7,
                token_modifiers_bitset: 128,
            },
        ),
        CC: Some(
            TokenBits {
                token_type: 16,
                token_modifiers_bitset: 4,
            },
        ),
        VBG: Some(
            TokenBits {
                token_type: 12,
                token_modifiers_bitset: 64,
            },
        ),
        WDT: Some(
            TokenBits {
                token_type: 15,
                token_modifiers_bitset: 256,
            },
        ),
        PRP: Some(
            TokenBits {
                token_type: 9,
                token_modifiers_bitset: 0,
            },
        ),
        JJS: Some(
            TokenBits {
                token_type: 4,
                token_modifiers_bitset: 512,
            },
        ),
        RBS: Some(
            TokenBits {
                token_type: 10,
                token_modifiers_bitset: 512,
            },
        ),
        O: Some(
            TokenBits {
                token_type: 17,
                token_modifiers_bitset: 16,
            },
        ),
        DT: Some(
            TokenBits {
                token_type: 18,
                token_modifiers_bitset: 256,
            },
        ),
        WRB: Some(
            TokenBits {
                token_type: 15,
                token_modifiers_bitset: 64,
            },
        ),
        EX: Some(
            TokenBits {
                token_type: 15,
                token_modifiers_bitset: 2,
            },
        ),
        VBP: Some(
            TokenBits {
                token_type: 12,
                token_modifiers_bitset: 4,
            },
        ),
        PDT: Some(
            TokenBits {
                token_type: 18,
                token_modifiers_bitset: 32,
            },
        ),
    },
)
